arch-x86: Move CPUID values to python

abmerop opened the pull request:
arch-x86: Move CPUID values to python

    CPUID values for X86 are currently hard-coded in the C++ source file.
    This makes it difficult to configure the bits if needed. Move these to
    python instead. This will provide a few benefits:

    1. We can enable features for certain configurations, for example AVX
    can be enabled when the KVM CPU is used, but otherwise should not be
    enabled as gem5 does not have full AVX support.
    2. We can more accurately communicate things like cache/TLB sizes based
    on the actual gem5 configuration. The CPUID values are can be used by
    some libraries, e.g., MPI, to query system topology.
    3. Enabling some bits breaks things in certain configurations and this
    can be prevented by configuring in python. For example, enabling AVX
    seems to currently be breaking SMP, meaning gem5 can only boot one CPU
    in that configuration.

abmerop committed:
arch-x86: Move CPUID values to python

CPUID values for X86 are currently hard-coded in the C++ source file.
This makes it difficult to configure the bits if needed. Move these to
python instead. This will provide a few benefits:

1. We can enable features for certain configurations, for example AVX
can be enabled when the KVM CPU is used, but otherwise should not be
enabled as gem5 does not have full AVX support.
2. We can more accurately communicate things like cache/TLB sizes based
on the actual gem5 configuration. The CPUID values are can be used by
some libraries, e.g., MPI, to query system topology.
3. Enabling some bits breaks things in certain configurations and this
can be prevented by configuring in python. For example, enabling AVX
seems to currently be breaking SMP, meaning gem5 can only boot one CPU
in that configuration.

Change-Id: Ib3866f39c86d61374b9451e60b119a3155575884

abmerop committed:
arch-x86: Support CPUID functions with indexes

Various CPUID functions will return different values depending on the
value of ECX when executing the CPUID instruction. Add support for this
in the X86 KVM CPU. A subsequent patch will add a CPUID function which
requires iterating through multiple ECX values.

Change-Id: Ib44a52be52ea632d5e2cee3fb2ca390b60a7202a

abmerop committed:
arch-x86: Expose CR4.osxsave bit

Related to the recent changes with moving CPUID values to python, this
value is needed to enable AVX and needs a way to be exposed to python as
well in order to set the bit and the corresponding CPUID values at the
same time.

Change-Id: I3cadb0fe61ff4ebf6de903018a8d8a411bfdb4e0

abmerop committed:
configs: Enable AVX2 for GPUFS+KVM

AVX is a requirement for some ROCm libraries, such as rocBLAS, which are
themselves requirements for libraries higher up the stack like PyTorch.
This patch sets the necessary CPUID bits in the GPUFS config to enable
AVX, AVX2, and various SSE features so that applications using these
libraries do not cause an illegal instruction trap.

Change-Id: Id22f543fb2a06b268271725a54075ee6a9a1f041

mattsinc commented on configs/example/gpufs/system/system.py:
```diff
@@ -231,6 +231,22 @@ def makeGpuFSSystem(args):
+        system.workload.enable_osxsave = 1
+        use_avx = True
+
+    avx_extended_state = [0x00000007, 0x00000340, 0x00000000, 0x00000340,
```
Is it possible to add a comment explaining what these numbers mean?

abmerop replied:
Yes I can do that-  these are related to how much storage the CPU tells the CPU has for xsave.

mattsinc commented on src/arch/x86/cpuid.cc:
```diff
@@ -31,162 +31,146 @@
-        APMInfo,
-        LongModeAddressSize,
-
-        /*
```
seems like these are still not implemented and should have been left?

abmerop replied:
The reason I decided to remove these is because there are dozens of missing functions so I don't think these ones are in any way special.  In a patch in this "rel change" (if we can still call it that), it sets an enum to a value which makes the compiler skip the other values. I am guessing the original author may have added these so that the final/last enum value returns the number of functions to the CPUID call correctly

mattsinc replied:
Ok ... I don't know this code well enough to say if this is a good idea or not, but I guess at least GitHub will document it here for future reference.

abmerop replied:
There are dozens of CPUID functions missing and I see no reason to define them here.  Since it's possible to just assign a number to the enum name, I suggest using that rather than having these placeholder functions that aren't actually implemented just to get the number to line up.

mattsinc commented on src/arch/x86/cpuid.hh:
```diff
@@ -38,28 +41,41 @@ class ThreadContext;
-                    uint64_t _rdx, uint64_t _rcx) :
-            rax(_rax), rbx(_rbx), rcx(_rcx), rdx(_rdx)
-        {}
+struct CpuidResult
```
seems you removed indentation here?

abmerop replied:
Yes that is true -  I made an "undintation" to this PR to be more in line with the namespace changes that occurred in gem5 v22 (around that time).

mattsinc reviewed (commented):
I don't feel 100% confident in some of these changes, but have at least a couple questions for you in the below.

abmerop commented:
Hmm, well I have to admit I think python black makes it look worse in this particular situation, but in an effort of keeping with the style guide I will re-upload.

powerjg commented on src/arch/x86/X86ISA.py:
```diff
@@ -54,3 +54,52 @@ class X86ISA(BaseISA):
+    #
+    # A good resource for these values can be found here:
+    #     https://sandpile.org/x86/cpuid.htm
+    # 0000_0001h
```
As Matt mentioned above, it would be good to add comments for what these default values mean.

I'm sure you did a lot of leg work to figure out all of these details. It would be great if we didn't have to do this again :)

abmerop replied:
These are just moved from the C++ and the values have not changed, so I did not really put in any work to understand them.  I put a link to a website above that explains the values.  It would be way too much to explain what all 1024 bits in this file do

powerjg commented on src/arch/x86/cpuid.hh:
```diff
@@ -38,28 +41,41 @@ class ThreadContext;
+class X86CPUID
+{
+  public:
+    X86CPUID(const X86ISAParams &p);
```
I don't like the idea of passing the SimObject params to a non-SimObject. I know that it will make things more verbose, but I think it would be better to explicitly pass in the parameters that you need.

I'm worried about the future if/when anything about params changes. We're coupling `X86CPUID` to `BaseParams` in a way that is unnecessary.

This isn't a big problem, though.

abmerop replied:
I originally tried to make it a SimObject but I could not figure out a way (if there is one) to create a default SimObject with all the default values. Therefore, every single python script which creates an X86 CPU would have to be changed, I think, which I wanted to avoid this way. I could potentially make the X86CPUID class in the X86ISA namespace/class.

powerjg replied:
The class not being a SimObject isn't a problem. The (small) issue I'm raising is that I don't like that a class that is not a SimObject is taking a SimObjectParams* as the parameter to its constructor.

I would  prefer something like
```
X86CPUID::X86CPUID(std::string vendor_string, std::string name_string)
    : vendorString(vendor_string), nameString(name_string)
{...}
```
and
```
cpuid = new X86CPUID(p.vendor_string, p.name_string);
```

This way `X86CPUID` is not coupled to `X86ISAParams` but only to `X86::ISA`

abmerop replied:
But the class is only used in X86ISA.  Doesn't your approach have the reserve problem where X86ISA depends on the enum value in X86CPUID?

powerjg replied:
This isn't a big deal, so you can feel free to ignore.

That's exactly right. Since X86CPUID is used only in X86ISA, it's OK for X86ISA to be coupled to X86CPUID. What we don't want to couple a class that's only used in one place (X86CPUID) to a class that is outside of the same "module" (i.e., X86ISAParams).

Again, not a big deal, but I think it would be better practice to not pass around the params object. Instead of passing around the params object, I would rather see passing the explicit variable/params from that object as parameters to the class.

powerjg commented on src/arch/x86/isa.hh:
```diff
@@ -93,6 +94,8 @@ class ISA : public BaseISA
 
     std::string getVendorString() const;
+
+    X86CPUID *cpuid;
```
Probably should be `unique_ptr`

powerjg replied:
I would like to see this change before merging.

powerjg replied:
Any reason not to make this a `unique_ptr`?

abmerop replied:
Missed this conversation somehow (kind find them hard to see in github's dark mode).  Done now.

powerjg commented:
> Hmm, well I have to admit I think python black makes it look worse in this particular situation, but in an effort of keeping with the style guide I will re-upload.

I think that it's worth having things that don't look perfect so we don't argue about style :)

andysan commented:
> > Hmm, well I have to admit I think python black makes it look worse in this particular situation, but in an effort of keeping with the style guide I will re-upload.
> 
> I think that it's worth having things that don't look perfect so we don't argue about style :)

There are cases where it might be legitimate to disable black. Black has a couple of [magic comments](https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html) (`# fmt: off/on`) that let you do this. I sometimes use this in my own code when formatting large tables where there is some structure to the data that I want to replicate in the source code. Another trick is to leave a hanging comma which tells black to format a dicts / lists / tuples / parameters with one entry per line.

BobbyRBruce commented:
The pre-commit checks failed. You can fix this locally using:

```sh
./util/pre-commit-install.sh # Ensure the hooks are installed
pre-commit run --files <files modified in this patch>
git add <files fixed by pre-commit>
git commit -m "misc: Fix pre-commit formatting issues"
```

Then push your changes.

abmerop committed:
arch-x86: Add extended state CPUID function

The extended state CPUID function is used to set the values of the XCR0
register as well as specify the size of storage for context switching
storage for x87 and AVX+. This function is iterative and therefore
requires (1) marking it as such in the hsaSignificantIndex function (2)
setting multiple sets of 4-tuples for the default CPUID values where the
last 4-tuple ends with all zeros.

Change-Id: Ib6a43925afb1cae75f61d8acff52a3cc26ce17c8

abmerop commented:
Not sure what I need to do for the PR list to show that this was updated, but it was updated.

powerjg reviewed (commented):
One more little thing. Let me know if you disagree with the suggestion. IMO, it's better to use smart pointers when possible and not use `new`/`delete`, but let me know if I'm missing something.

Thanks for the changes so far!

powerjg commented:
> Not sure what I need to do for the PR list to show that this was updated, but it was updated.

I think you can click the "re-review" button in the upper right by my name under "reviewers" to notify me. 

![image](https://github.com/gem5/gem5/assets/109596/ed88d8c0-7a70-4438-9e32-d28298b8fe66)

powerjg commented:
(Apologies for closing, my keyboard did something crazy...)

powerjg reviewed (dismissed):
Thanks for all of the updates!

BobbyRBruce commented:
The tests appear to be failing to run the `test-x86-hello64-dynamic` tests. Since the contributors to this patch are currently sleeping i'll return the tests just to make sure this isn't a bug with our runners.

abmerop commented:
> The tests appear to be failing to run the `test-x86-hello64-dynamic` tests. Since the contributors to this patch are currently sleeping i'll return the tests just to make sure this isn't a bug with our runners.

Seems they are still failing.  I have a few questions about this:  How can I download the artifacts(?) that show what the error is?  I can see which tests failed, how can I run *only* those tests (using ./tests/main.py?).  I wasn't sure how to do this so I ran the whole thing overnight (./main run -vv) but it appears to have either hanged or takes longer than 12 hours.

Edit: More specifically I ran ./main.py run -j60 -t60 -vv.

abmerop commented:
> > The tests appear to be failing to run the `test-x86-hello64-dynamic` tests. Since the contributors to this patch are currently sleeping i'll return the tests just to make sure this isn't a bug with our runners.
> 
> Seems they are still failing. I have a few questions about this: How can I download the artifacts(?) that show what the error is? I can see which tests failed, how can I run _only_ those tests (using ./tests/main.py?). I wasn't sure how to do this so I ran the whole thing overnight (./main run -vv) but it appears to have either hanged or takes longer than 12 hours.
> 
> Edit: More specifically I ran ./main.py run -j60 -t60 -vv.

Was able to find the command with -vvv.  I ran the three failing tests locally and they all pass (the amd64 hello world apps).  This makes sense to me as hello world... doesn't use CPUID probably.

powerjg commented:
For next time, you can download the results from the test on the actions page:
![image](https://github.com/gem5/gem5/assets/109596/083c3c69-1446-4ed1-99fa-902c58c016eb)

Then, you can search for the failing test and find the simout/simerr, which also includes the command that was run.

In this case, simerr shows the problem:
```
src/base/statistics.hh:279: warn: One of the stats is a legacy stat. Legacy stat is a stat that does not belong to any statistics::Group. Legacy stat is deprecated.
src/base/remote_gdb.cc:418: warn: Sockets disabled, not accepting gdb connections
src/sim/simulate.cc:194: info: Entering event queue @ 0.  Starting simulation...
src/sim/mem_state.cc:443: info: Increasing stack size by one page.
src/sim/syscall_emul.cc:74: warn: ignoring syscall set_robust_list(...)
src/sim/syscall_emul.cc:74: warn: ignoring syscall rseq(...)
/lib/x86_64-linux-gnu/libc.so.6: CPU ISA level is lower than required
```

It looks like the changes to CPUID is causing glibc on the host (docker container running gem5 in this case) to get confused as to the CPU level.

The exact command (from the simout file) is:
```
command line: /__w/gem5/gem5/build/ALL/gem5.opt -d /tmp/gem5outx1tcuelx -re --silent-redirect /__w/gem5/gem5/tests/gem5/configs/simple_binary_run.py x86-hello64-dynamic timing --resource-directory /__w/gem5/gem5/tests/gem5/resources x86
```

Of course, you can ignore the paths, etc. when trying to reproduce.

Finally, the docker container that we are using is `gcr.io/gem5-test/ubuntu-22.04_all-dependencies:latest` as found here: https://github.com/gem5/gem5/actions/runs/5686873518/workflow?pr=113#L78

Let me know if there's more we can do to help you diagnose this in the future :smile:

powerjg commented:
(Side note: I want to point out how much better the situation is on github for diagnosing when tests fail! We now have access/control over every aspect of the testing! Rather than having to re-run all of the tests because one little thing failed for an unknown reason (like when we were using kokoro), we can now see the *output* of the failing test and easily reproduce!)

abmerop commented:
> (Side note: I want to point out how much better the situation is on github for diagnosing when tests fail! We now have access/control over every aspect of the testing! Rather than having to re-run all of the tests because one little thing failed for an unknown reason (like when we were using kokoro), we can now see the _output_ of the failing test and easily reproduce!)

Thanks Jason.  I'm sure everything in here ... somewhere.  So far I am finding the github interface is much more complex than gerrit and not really able to find anything since i've never used github much.

abmerop commented:
> For next time, you can download the results from the test on the actions page: ![image](https://user-images.githubusercontent.com/109596/256862702-083c3c69-1446-4ed1-99fa-902c58c016eb.png)
> 
> Then, you can search for the failing test and find the simout/simerr, which also includes the command that was run.
> 
> In this case, simerr shows the problem:
> 
> ```
> src/base/statistics.hh:279: warn: One of the stats is a legacy stat. Legacy stat is a stat that does not belong to any statistics::Group. Legacy stat is deprecated.
> src/base/remote_gdb.cc:418: warn: Sockets disabled, not accepting gdb connections
> src/sim/simulate.cc:194: info: Entering event queue @ 0.  Starting simulation...
> src/sim/mem_state.cc:443: info: Increasing stack size by one page.
> src/sim/syscall_emul.cc:74: warn: ignoring syscall set_robust_list(...)
> src/sim/syscall_emul.cc:74: warn: ignoring syscall rseq(...)
> /lib/x86_64-linux-gnu/libc.so.6: CPU ISA level is lower than required
> ```
> 
> It looks like the changes to CPUID is causing glibc on the host (docker container running gem5 in this case) to get confused as to the CPU level.
> 
> The exact command (from the simout file) is:
> 
> ```
> command line: /__w/gem5/gem5/build/ALL/gem5.opt -d /tmp/gem5outx1tcuelx -re --silent-redirect /__w/gem5/gem5/tests/gem5/configs/simple_binary_run.py x86-hello64-dynamic timing --resource-directory /__w/gem5/gem5/tests/gem5/resources x86
> ```
> 
> Of course, you can ignore the paths, etc. when trying to reproduce.
> 
> Finally, the docker container that we are using is `gcr.io/gem5-test/ubuntu-22.04_all-dependencies:latest` as found here: https://github.com/gem5/gem5/actions/runs/5686873518/workflow?pr=113#L78
> 
> Let me know if there's more we can do to help you diagnose this in the future 😄

I guess I will need to run on Ubuntu 22 to see this error then.

powerjg commented:
You can just use docker.

```
docker run --rm -w `pwd` -v `pwd`:`pwd` -u $UID:$GID gcr.io/gem5-test/ubuntu-22.04_all-dependencies:latest build/ALL/gem5.opt tests/gem5/configs/simple_binary_run.py x86-hello64-dynamic timing x86
```

I *think* that command works (I didn't test it)

abmerop commented:
Fixed the tester issue
